<h1 align="center">ðŸ§¬WildSEQ:</h1>
<h3 align="center">A command-line analysis tool for species identification using  qPCR 
sequencing data</h3>


## INSTALLATION
**Miniforge3** is recommended as this isnâ€™t affected by the Anaconda licensing issues for commercial 
sites. A script is provided that will set up your miniforge3 environment if you havenâ€™t already got 
one. Just run the `/installers/install_miniforge3.sh` script and it will set up your environment. 
Once it completes, close your current shell and open a fresh one so the conda command becomes 
available. 

If you have miniforge3 available already or you have run the included install script, you 
can run `/installers/install_wildseq_env.sh` to create a conda env for the script to use or manually 
install making sure you have all the package versions as specified in this script or as listed in
`/scripts/wildseq_dependencies.tsv`. 

**Donâ€™t forget to activate the conda env before running the WILDSEQ script:** `conda activate wildseq`.

## SYNOPSIS
**WildSEQ** is based on [LAMPSEQ](https://github.com/oliviayugovich/LAMPSEQ), adapted for qPCR data. In brief, **WildSEQ** takes input `fastq` files and performs quality control, primer trimming, alignment, and 
species identification, and outputs the relevant results which can be found in `/indir/05_results/` 
including:
- `indir_multiqc.html`: quality control run summary from `MultiQC`;
- `indir_species_calls.csv`: csv file containing `barcode`, `species_ID` and `reads` which can be used 
for user-defined plotting or further analysis;
- `WildSEQ_indir_run_summary.pdf`: a run summary pdf file including a run overview, per barcode plots, and summary table. 

To run the script: 
- Follow the installation instructions
- Ensure basecalled and demultiplexed `fastq` files to be analysed are located in: `/indir/00_raw_data/`
- Then run the following: `WILDSEQ.sh â€“-indir [input directory]`

## OPTIONS
<table> <colgroup> <col style="width: 22%" /> <col style="width: 77%" /> </colgroup> <thead> <tr 
class="header"> <th>Option</th> <th>Description</th> </tr> </thead> <tbody> <tr class="odd"> 
<td><code>--indir &lt;value&gt;</code></td> <td>Set input directory (defaults to current working 
directory)</td> </tr> <tr class="even"> <td><code>--dryrun</code></td> <td>Set dryrun flag for testing 
without running (default=false)</td> </tr> <tr class="odd"> <td><code>--debug</code></td> <td>Set 
debug flag for verbose output (default=false unless dryrun set)</td> </tr> <tr class="even"> 
<td><code>--help</code></td> <td>Print this message</td> </tr> </tbody> </table>

## TEST DATA
To check the script is working as intended, you can run `WILDSEQ.sh` on the `/test_data/` directory. 
The relevant `fastq` files can be found in `/test_data/00_raw_data/`. To analyse the `/test_data/`, 
run the following: `WILDSEQ.sh --indir test_data`. 

Once the analysis is complete, cross-check that the output files generated by the code in `/test_data/05_results/` match those found in `/test_data/06_test_data_answers/`. 

A clean up script has been provided to quickly remove output files from the `/test_data/` directory. To run this, execute `/test_data/clean_test.sh`.

**The test data provided is a subset of data generated for our study** *(see citation below)*. For details on how the data was produced, please refer to the **Methods section** of the paper. Some relevant supporting files are included in this repository:
- **Sample and barcode information:**
  See `/test_data/test_data_analysis_sheet.csv`
- **Reference sequence GenBank accessions:**
  See `/scripts/ref_seqs/ref_seqs_source.csv` or in the **Methods section** of the paper.
- **qPCR primer sequences:**
  See **Methods section** of the paper.


## DEPENDENCIES
For dependencies see `/scripts/wildseq_dependencies.tsv`.

## CITATION
**This tool has been developed by Olivia Yugovich and Shane Sturrock.** If you use **WildSEQ** in your work, please cite the following manuscript:
> Yugovich, O., Cave, V., Sturrock, S., & Harbison, S.A. (2026). *Morphological-based preservation scoring predicts DNA detection success in elephant and mammoth ivory, bone, and teeth using WildSEQ* Submitted and currently under review at *Forensic Science International: Genetics*.


## CONTACT
To contact the authors about this code or customising for other qPCR targets, **please email Olivia Yugovich** at olivia.yugovich@gmail.com
